### ISSUE TITLE
mypy incorrectly infers callable signature when assigning async function to class attribute

### ISSUE THREAD

**Issue by wooparadog** [python/mypy#19713]
**Bug Report**

**Bug Description:**
mypy incorrectly infers the callable type when assigning an async function to a class attribute, losing the parameter information and causing false positive type errors.

**To Reproduce:**

```python
from abc import ABC
from typing import Any, Callable, Coroutine

async def function(arg) -> bool:
    print(arg)
    return True

class BaseClass(ABC):
    class_member_function: Callable[[Any], Coroutine[Any, Any, bool]] | None = None

class B(BaseClass):
    class_member_function = function  # Error here

import asyncio
asyncio.run(B.class_member_function("!!"))
```

**Actual Behavior:**
```
error: Incompatible types in assignment (expression has type "Callable[[], Coroutine[Any, Any, bool]]", base class "BaseClass" defined the type as "Callable[[Any], Coroutine[Any, Any, bool]] | None")  [assignment]
```

**Expected Behavior:**
No type error should occur. The `function` has signature `(arg) -> bool` which should be compatible with `Callable[[Any], Coroutine[Any, Any, bool]]`.

**Root Cause:**
mypy incorrectly infers `function` as `Callable[[], Coroutine[Any, Any, bool]]` (no parameters) instead of `Callable[[Any], Coroutine[Any, Any, bool]]` (one parameter) when used in class attribute assignment.

**Environment:**
- mypy version: 1.17.1
- Python version: 3.x

**Workaround:**
Explicit type annotation works:
```python
class B(BaseClass):
    class_member_function: Callable[[Any], Coroutine[Any, Any, bool]] = function
```

**Comment by wooparadog** [comment]
**Update: Workaround has limitations**

I tested the workaround with explicit type annotation using pyright, and it reveals another issue:

```
/test.py:15:5 - error: "class_member_function" overrides symbol of same name in class "BaseClass"
  Variable is mutable so its type is invariant
    Override type "(Any) -> Coroutine[Any, Any, bool]" is not the same as base type "((Any) -> Coroutine[Any, Any, bool]) | None" (reportIncompatibleVariableOverride)
```

So while the explicit type annotation works around the mypy inference bug, it creates a new issue with pyright due to the `| None` union type in the base class. This suggests the original mypy inference bug is more problematic than initially apparent, as it prevents proper type-safe inheritance patterns.

**Environment:**
- pyright 1.1.390

**Comment by A5rocks** [comment]
This is not async specific:
```py
from abc import ABC
from typing import Any, Callable

def function(arg) -> bool:
    print(arg)
    return True

class BaseClass(ABC):
    class_member_function: Callable[[Any], bool] | None = None

class B(BaseClass):
    class_member_function = function  # Error here

B.class_member_function("!!")
```

IIRC there's an issue about this already but I can't find it.

**Comment by wooparadog** [comment]
> IIRC there's an issue about this already but I can't find it.

Yeah, I've also searched before hand. Couldn't find it either.

**Comment by sterliakov** [comment]
I'm not sure this is actually a bug?

`mypy` recognizes this assignment as a method assignment, binding first argument to `self`. If the function were typed, you'd also get "incompatible self" errors. The problem is that given

```python
class Foo:
    fn: Callable[[Any], None]
```

the following should be valid:

```python
def run(foo: Foo) -> None:
    foo.fn(0)
```

and it is not in either of your examples. 


To make access from class work, you can wrap the declaration in `ClassVar`:

```python
from abc import ABC
from typing import Any, Callable, ClassVar, Coroutine

async def function(arg) -> bool:
    print(arg)
    return True

class BaseClass(ABC):
    class_member_function: ClassVar[Callable[[Any], Coroutine[Any, Any, bool]] | None] = None

class B(BaseClass):
    class_member_function = function  # Error here

import asyncio
asyncio.run(B.class_member_function("!!"))
asyncio.run(B().class_member_function())
asyncio.run(B().class_member_function(''))  # E: too many arguments
```

Yes, `Callable` and `ClassVar[Callable]` annotations affect the method binding as understood by `mypy`. 

In your example mypy correctly rejects assignment after method binding, because `class_member_function` has signature `() -> bool` when accessed on `B` instance, conflicting with `(Any) -> bool` definition in superclass.

**Comment by wooparadog** [comment]
Thanks. It's indeed better to explicitly define the function as `ClassVar`. Issue closed.
